Internationalization with `i18next`

demize opened the pull request:
### Checklist

- [X] Have you followed the guidelines in our Contributing document?
- [X] Have you run `eslint` on the code and resolved any errors?
- [X] Have you run `npm test` and all tests are passing?
- [ ] Have you added new tests for any new functions created?
- [X] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?
- [X] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you updated CHANGELOG.md to add your changed under the `[Unreleased]` heading?

### Description

This implements internationalization with the `i18next` library, and provides an English locale and a pseudo-localized `dev` locale. 

### Benefits

More people might be able to use the bot if it's available in their language.

### Potential drawbacks

This is harder to maintain. The compiled bot is larger, since `i18next` is a large library. We don't have any actual translations done yet, so this change has little benefit yet. People might start using the `dev` locale because they think it looks cool...

Additionally, not every string goes through i18next, and some of them are built in ways that are probably not conducive to actual localization effort. I've done what I can to get everything I think needs to be internationalized switched over, but we're probably not going to have a perfect solution (at least not without feedback from people who actually try to localize the bot).

demize commented:
I think I've got everything we want to localize migrated over now. The only remaining strings are more useful for debugging than anything else, so we probably don't want to localize them; migrating them over is easy enough if we need to, though.

liquidnya commented on src/queue.ts:
```diff
@@ -1124,10 +1128,28 @@ const queue = {
-      return result;
+
+      if (weightedList.entries.length > 5) {
+        onlineList += "etc.";
```
I would add a translation for this word. For example `etc.` is `usw.` in German.
Maybe we could even give the `onlineList` as the argument of the translation and then have an etc. version and a version without etc.?

liquidnya commented on src/queue.ts:
```diff
@@ -1124,10 +1128,28 @@ const queue = {
             "%)",
           ""
         );
```
We should probably use `i18next` for the list of levels for 2 reasons:
Some languages can use different brackets like Japanese for example.
Some languages can use different separators of items:
https://linguistics.stackexchange.com/questions/6737/what-languages-use-the-serial-comma#answer-6791

Maybe we should just use `Intl.ListFormat` for formatting lists in general (also for the other places where lists are printed like custom codes and custom levels).
See https://www.i18next.com/translation-function/formatting#list
And then also add a translation entry for the `{{submitter}} ({{percentage}}%)` or something like this.

liquidnya reviewed (changes requested):
Awesome work!
Just the 2 instances of the word `etc.` need some translations and we should print lists differently.

liquidnya commented:
The function `queue.percent` in `src/queue.ts` would need to be changed as well.
`percent.toFixed` doesn't return a localized percentage.
For example in German `100.0` would be `100,0`.

liquidnya commented on src/index.ts:
```diff
@@ -203,38 +170,34 @@ const weightedchance_message = async (
+    return i18next.t("senderOddsMultiplier", {
+      sender,
+      chance,
+      multiplier: multiplier.toFixed(1),
```
Here is another usage of `toFixed` which would need to be changed.
Numbers could be formatted like this: https://www.i18next.com/translation-function/formatting#passing-options-to-the-formatting using the `minimumFractionDigits`.

demize commented on src/extensions.ts:
```diff
@@ -35,13 +35,14 @@ import {
 // internal interface
 interface ExtensionModule<T = Promise<void> | void> {
-  setup(api: ExtensionsApi): T;
+  setup(api: ExtensionsApi, i18next: i18n): T;
```
No, this is an artifact of my attempts to get i18next working with extensions. I've removed it, as that attempt was unsuccessful.

demize commented:
@liquidnya Second-to-last commit I think addresses your concerns, though perhaps more radically than you expected :p

Honestly I think I like this approach better (no more ">99.9%" and "<0.01%", instead we get slightly more precise numbers on the low end and a tendency not to round up to 100) both because it's a lot easier to make with with localization and because it's just... aesthetically nicer.

liquidnya commented:
I really liked how the queue printed `>99.9%` before; but I do like this solution better than the `<0.01%`, something like `0.0414%` is way nicer!
I guess it printing `100%` is very unlikely.
I tested with `x/(x+1)`, and the first time it prints `100%` is when `x=2000`, so when someone waited 33hours and 20min and then someone else adds a level it will display `100%` ^^
So it sounds very very unlikely to me anyways.

demize commented:
`1999/2000` is 0.9995, which is the half-increment (exactly 99.95%), so anything after that is going to be rounded up with `halfTrunc`; I think that by that point, we might as well just say 100%, the chances are so astronomical of both getting there *and* of it picking anyone else in the queue that 100% is probably pretty accurate anyway 😅.

liquidnya commented on locales/en/fluid-queue.json:
```diff
@@ -0,0 +1,100 @@
+  "persistenceReloaded": "Reloaded queue state from disk.",
+  "persistenceInvalid": "Invalid arguments. The correct syntax is !persistence {on/off/save/load}.",
+  "queueEmpty": "There are no $t(level, {\"count\": 0}) in the queue.",
+  "queueListCurrent": "{{onlineLength}} online: {{- current.submitter}} (current), {{- onlineList, list}}... ({{levels.offline.length}} offline)",
+  "queueListCurrentOnly": "{{onlineLength}} online: {{- current.submitter}} (current)... ({{levels.offline.length}} offline)",
+  "queueListNoCurrent": "{{onlineLength}} online: (no current level), {{- onlineList, list}}... ({{levels.offline.length}} offline)",
+  "submitterPercentage": "{{submitter}} ({{percent, number}})",
+  "genericPercentage": "{{percent, number}}",
+  "weightedListCurrent": "{{onlineLength}} online: {{- current.submitter}} (current), {{- onlineList, list}}... ({{weightedList.offlineLength}} offline)",
+  "weightedListCurrentOnly": "{{onlineLength}} online: {{- current.submitter}} (current)... ({{weightedList.offlineLength}} offline)",
+  "weightedListNoCurrent": "{{onlineLength}} online: (no current level), {{- onlineList, list}}... ({{weightedList.offlineLength}} offline)",
```
I see you have removed the `etc.`.
There are still those `...` which I know have been here always, but I think we do not need those `...`, because there is no semantic to it. (It doesn't even say that there are more levels in the queue than the ones displayed, because it there are 2 levels in the queue it will display both of those levels and then still those dots)

I think we should remove the dots and maybe there could be a way to tell that there are more levels following (levels.online.length > 5) which could be the dots or maybe an `etc.`.

demize replied:
Ah, I forgot to mention the `etc`...

And it's probably good I did. I wrote out a bunch here as to why I got rid of it, but it got me to think about it more, and now there's still no `etc` but I've updated it to conditionally print an ellipsis through some (what I think is clever) nesting of localization strings. I've also added a test to test the ellipses, with some very unimaginative usernames and some random level codes picked from my Discord.

liquidnya commented on src/index.ts:
```diff
@@ -9,6 +9,8 @@ import settings from "./settings.js";
 import { helper } from "./chatbot.js";
 import { QueueEntry } from "./extensions-api/queue-entry.js";
 import { Chatter, Responder } from "./extensions-api/command.js";
+import "./i18n.js";
```
I would like this to be `await import("./i18n.js");`, because otherwise
```
Initializing i18next...
Language: en
```
would be printed above the banner.

demize replied:
Good catch. I also didn't mean to keep printing the language there, so I've removed that.

liquidnya commented on src/index.ts:
```diff
@@ -790,26 +719,26 @@ async function HandleMessage(
-      respond("Next level order: " + order.reduce((acc, x) => acc + ", " + x));
+      respond(
+        i18next.t("orderList", {
+          list: order.reduce((acc, x) => acc + ", " + x),
```
Here we could pass the list instead of using `", "`.

liquidnya commented on src/queue.ts:
```diff
@@ -531,21 +532,31 @@ const queue = {
+            list: weightedList.entries
               .map((entry) => entry.level.submitter.toString())
-              .reduce((a, b) => a + ", " + b)
+              .reduce((a, b) => a + ", " + b),
```
Here we could pass the list instead of using `", "`.

liquidnya commented on src/queue.ts:
```diff
@@ -728,30 +739,31 @@ const queue = {
+          list: weightedList.entries
             .map((entry) => entry.level.submitter.toString())
-            .reduce((a, b) => a + ", " + b)
+            .reduce((a, b) => a + ", " + b),
```
Here we could pass the list instead of using `", "`.

liquidnya commented on src/extensions.ts:
```diff
@@ -244,7 +245,11 @@ export class Extensions {
-    console.log(`Extensions: [${Object.keys(this.extensions).join(", ")}]`);
+    console.log(
+      i18next.t("extensionsList", {
+        extensions: Object.keys(this.extensions).join(", "),
```
We could translate the list instead of using `.join(", ")`.

liquidnya commented on src/extensions/customcode.ts:
```diff
@@ -105,10 +111,10 @@ const commandHandler = (
       } else {
-        const response = list.join(", ");
-        return "The current custom codes are: " + response + ".";
+        const codes = list.join(", ");
```
We could translate the list instead of using `.join(", ")`.

liquidnya commented on src/extensions/customlevel.ts:
```diff
@@ -250,9 +256,12 @@ const customlevelCommand = (custom: CustomData) => {
+                    ns: "customlevel",
+                    code,
+                    name: custom.get(fromCode).name,
+                    codes: custom.get(fromCode).codes.join(", "),
```
We could translate the list instead of using `.join(", ")`.

liquidnya commented on src/extensions/customlevel.ts:
```diff
@@ -263,13 +272,21 @@ const customlevelCommand = (custom: CustomData) => {
+                    ns: "customlevel",
+                    code,
+                    name: value.name,
+                    codes: value.codes.join(", "),
```
We could translate the list instead of using `.join(", ")`.

liquidnya commented on src/extensions/customlevel.ts:
```diff
@@ -293,16 +311,24 @@ const customlevelCommand = (custom: CustomData) => {
+                    ns: "customlevel",
+                    code,
+                    name: custom.get(fromCode).name,
+                    codes: custom.get(fromCode).codes.join(", "),
```
We could translate the list instead of using `.join(", ")`.

liquidnya commented on src/extensions/customlevel.ts:
```diff
@@ -316,15 +342,31 @@ const customlevelCommand = (custom: CustomData) => {
+                  i18next.t("customlevelEnabled", {
+                    ns: "customlevel",
+                    name: value.name,
+                    codes: value.codes.join(", "),
```
We could translate the list instead of using `.join(", ")`.

liquidnya commented on src/extensions/customlevel.ts:
```diff
@@ -334,13 +376,20 @@ const customlevelCommand = (custom: CustomData) => {
+                i18next.t("customlevelRemoved", {
+                  ns: "customlevel",
+                  name: value.name,
+                  codes: value.codes.join(", "),
```
We could translate the list instead of using `.join(", ")`.

liquidnya commented on src/extensions/customlevel.ts:
```diff
@@ -352,28 +401,36 @@ const customlevelCommand = (custom: CustomData) => {
+                  ns: "customlevel",
+                  code,
+                  name: custom.get(fromCode).name,
+                  codes: custom.get(fromCode).codes.join(", "),
```
We could translate the list instead of using `.join(", ")`.

liquidnya commented on src/extensions/customlevel.ts:
```diff
@@ -410,42 +473,48 @@ const customlevelCommand = (custom: CustomData) => {
+                i18next.t("customlevelExists", {
+                  ns: "customlevel",
+                  name: custom.get(fromName).name,
+                  codes: custom.get(fromName).codes.join(", "),
```
We could translate the list instead of using `.join(", ")`.

liquidnya commented on src/extensions/customlevel.ts:
```diff
@@ -410,42 +473,48 @@ const customlevelCommand = (custom: CustomData) => {
+                    ns: "customlevel",
+                    code,
+                    name: custom.get(fromCode).name,
+                    codes: custom.get(fromCode).codes.join(", "),
```
We could translate the list instead of using `.join(", ")`.

liquidnya commented on src/extensions/customlevel.ts:
```diff
@@ -410,42 +473,48 @@ const customlevelCommand = (custom: CustomData) => {
+                  ns: "customlevel",
+                  uuid,
+                  name: custom.get(uuid).name,
+                  codes: custom.get(uuid).codes.join(", "),
```
We could translate the list instead of using `.join(", ")`.

liquidnya commented on src/extensions/customlevel.ts:
```diff
@@ -456,13 +525,15 @@ const customlevelCommand = (custom: CustomData) => {
+              i18next.t("customlevelAddedCodes", {
+                ns: "customlevel",
+                levelName,
+                codes: codes.join(", "),
```
We could translate the list instead of using `.join(", ")`.

liquidnya reviewed (changes requested):
I found some further places where we could translate a list (and also translate brackets).

demize commented:
I have to stream now, but I'll go through and address all the lists we can be printing better after stream :P

demize commented:
Okay, that should be all the joins and reduces addressed...

liquidnya reviewed (changes requested):
I forgot to mention this place here in the last review 😅
https://github.com/fluid-queue/fluid-queue/blob/f72f257b894aa5bbaf00a1ccc427934346d4de92/src/extensions/customlevel.ts#L159-L162
This code has a list and also uses brackets.

https://github.com/fluid-queue/fluid-queue/blob/f72f257b894aa5bbaf00a1ccc427934346d4de92/src/extensions-api/resolvers.ts#L369

This is also a list and misses translation as well (but maybe not translating this is on purpose).

liquidnya reviewed (approved):
LGTM
